---
title: "Metagenomic ORF Detection and Annotation"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(cache = FALSE)
options(scipen=999)
cpal=4

library(data.table,quietly = T)
library(reshape2,quietly = T)
library(ggplot2,quietly = T)
library(RColorBrewer,quietly = T)
library(cowplot,quietly = T)
library(stringr,quietly = T)
theme_set(theme_cowplot())

#read in the ORF summary table
orf=fread(commandArgs(trailingOnly=T)[2],header = T,sep="\t")

#read in the gtf summary table
gtf=fread(commandArgs(trailingOnly=T)[1],header = T,sep="\t")

```

## ORF Detection

```{r, echo=FALSE, warning=FALSE}

noorfs=orf[nrow(orf),2]
meanlen=orf[nrow(orf),3]
sdlen=orf[nrow(orf),4]
meanno=orf[nrow(orf),5]
sdno=orf[nrow(orf),6]
#remove totals row
orf=orf[-nrow(orf),]

```

A total of **`r noorfs`** ORFs with length **`r meanlen` $\pm$ `r sdlen`** (Mean $\pm$ SD)  were detected across all samples.

There were **`r meanno` $\pm$ `r sdno`** (Mean $\pm$ SD) ORFs per contig.

Per sample ORF stats are given below (bars are means with SD as error bars):

```{r, echo=FALSE, warning=FALSE}
orf$Sample=str_match(orf$Sample,"Annotation/combined_annotations.dir/(\\S+).orf_annotations.gtf.gz")[,2]

noorfspersamp=ggplot(orf,aes(y=No_ORFs,x=Sample,fill=Sample))+geom_bar(stat="identity")+coord_flip()+xlab("Sample")+ylab("No. ORFs")+guides(fill=FALSE)
plot(noorfspersamp)

orfsizepersamp=ggplot(orf,aes(x=Sample,y=Mean_ORF_Size,fill=Sample))+geom_bar(stat="identity")+geom_errorbar(aes(ymin=Mean_ORF_Size-SD_ORF_Size,ymax=Mean_ORF_Size+SD_ORF_Size))+coord_flip()+ylab("ORF Size")+guides(fill=FALSE)
plot(orfsizepersamp)

orfspercont=ggplot(orf,aes(x=Sample,y=Mean_ORFS_Per_Contig,fill=Sample))+geom_bar(stat="identity")+geom_errorbar(aes(ymin=Mean_ORFS_Per_Contig-SD_ORFS_Per_Contig,ymax=Mean_ORFS_Per_Contig+SD_ORFS_Per_Contig))+coord_flip()+ylab("ORFs Per Contig")+guides(fill=FALSE)
plot(orfspercont)

```

````{r, echo=FALSE}
taxfeats=c("kingdom","phylum","class","order","family","genus","species")
tax=gtf[which(gtf$feat_type%in%taxfeats),]

fun=gtf[-which(gtf$feat_type%in%taxfeats),]

```

## Taxonomic Annotation

```{r, echo=FALSE}
for(i in taxfeats){
  cat(paste0(length(unique(tax$feat_name[tax$feat_type==i]))," unique ",i," observed across all samples.\n"))  
}

sampcnt=rowSums(tax[,3:ncol(tax)])
sampcnt=data.frame(feat=tax$feat_name,Taxon_Lev=tax$feat_type,Count=sampcnt)
scplot=ggplot(sampcnt,aes(x=Taxon_Lev,y=Count,fill=Taxon_Lev))+geom_boxplot()+xlab("Taxonomic Level")+ylab("No. Samples Taxon Observed In")+guides(fill=FALSE)+coord_flip()
plot(scplot)

```

## Functional Annotation
 
```{r, echo=FALSE}

for(i in unique(fun$feat_type)){
  cat(paste0(length(unique(fun$feat_name[fun$feat_type==i]))," unique ",i," observed across all samples.\n"))  
}

fsampcnt=rowSums(fun[,3:ncol(fun)])
fsampcnt=data.frame(feat=fun$feat_name,Fun_Type=fun$feat_type,Count=fsampcnt)
fscplot=ggplot(fsampcnt,aes(x=Fun_Type,y=Count,fill=Fun_Type))+geom_boxplot()+xlab("Functional Annotation Type")+ylab("No. Samples Annotation Observed In")+guides(fill=FALSE)+coord_flip()
plot(fscplot)

```
